google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterOperation.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.core.ApiFunction;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.util.concurrent.Executor;
+
+/**
+ * Represents a single write for BulkWriter, encapsulating operation dispatch and error handling.
+ */
+class BulkWriterOperation {
+  private final SettableApiFuture<WriteResult> future = SettableApiFuture.create();
```

Comment:
optional: rename to `operationFuture` to avoid confusion with the local future variables in the error and success handlers (or maybe it's just me that gets confused).